File: tugasFariz.py
# Deklarasi Fixed-Size Array kapasitas 10
# Di Python, Saya simulasikan fixed size dengan mengisi None sebanyak 10 kali
capacity = 10
nim_array = [None] * capacity
nama_array = [None] * capacity
jk_array = [None] * capacity

count = 0

# HEADER IDENTITAS
print("\n" + "="*49)
print("            TUGAS STRUKTUR DATA ARRAY            ")
print("="*49)
print(" Nama    : Fariz Husain Albar")
print(" NIM     : 24106050011")
print(" Kelas   : Informatika A")
print(" Matkul  : Struktur Data")
print("="*49)
input(" >> Tekan ENTER untuk masuk ke Menu Program...")

while True:
    print(f"\n=== MENU MAHASISWA (Count: {count}/{capacity}) ===")
    print("1. Insert at beginning")
    print("2. Insert at given position")
    print("3. Insert at end")
    print("4. Delete from beginning")
    print("5. Delete given position")
    print("6. Delete from end")
    print("7. Delete first occurence (by NIM)")
    print("8. Show data")
    print("9. Exit")

    try:
        # Input menu
        choice_input = input("Pilih menu: ")
        # Validasi apakah input angka atau bukan
        if not choice_input.isdigit():
             print(">> Input salah! Harap masukkan angka.")
             continue
        choice = int(choice_input)

    except ValueError:
        print(">> Input salah! Harap masukkan angka.")
        continue

    # LOGIKA MENU

    # 1. INSERT AT BEGINNING
    if choice == 1:
        if count >= capacity:
            print(">> GAGAL: Array Penuh!")
        else:
            # INPUT NIM (LOOP SAMPAI BENAR)
            nim = ""
            while True:
                nim = input("Masukkan NIM: ")

                # Cek apakah angka
                if not nim.isdigit():
                    print(">> ERROR: NIM harus berupa ANGKA saja! Coba lagi.")
                    continue

                is_duplicate = False
                for i in range(count):
                    if nim_array[i] == nim:
                        is_duplicate = True
                        break

                if is_duplicate:
                    print(f">> ERROR: NIM {nim} sudah ada! Coba NIM lain.")
                else:
                    break # Lolos Validasi

            # INPUT NAMA (LOOP SAMPAI BENAR)
            nama = ""
            while True:
                nama = input("Masukkan Nama: ")

                is_nama_duplicate = False
                for i in range(count):
                    # .lower() agar tidak case sensitive
                    if nama_array[i].lower() == nama.lower():
                        is_nama_duplicate = True
                        break

                if is_nama_duplicate:
                    print(f">> ERROR: Nama \"{nama}\" sudah ada! Coba nama lain.")
                else:
                    break

            # INPUT JK (LOOP SAMPAI BENAR)
            jk = ""
            while True:
                input_jk = input("Masukkan Jenis Kelamin (L/P): ").upper()
                if input_jk == "L" or input_jk == "P":
                    jk = input_jk
                    break
                else:
                    print(">> ERROR: Input salah! Ketik 'L' atau 'P'.")

            # PROSES INSERT DEPAN (Geser ke Kanan)
            # range(start, stop, step) -> mundur dari count ke 1
            for i in range(count, 0, -1):
                nim_array[i] = nim_array[i-1]
                nama_array[i] = nama_array[i-1]
                jk_array[i] = jk_array[i-1]

            nim_array[0] = nim
            nama_array[0] = nama
            jk_array[0] = jk
            count += 1
            print(">> SUKSES: Data disimpan di awal.")

    # 2. INSERT AT GIVEN POSITION
    elif choice == 2:
        if count >= capacity:
            print(">> GAGAL: Array Penuh!")
        else:
            # INPUT POSISI (LOOP SAMPAI BENAR)
            pos = 0
            while True:
                try:
                    pos_input = input(f"Masukkan Posisi Insert (1 - {count + 1}): ")
                    if not pos_input.isdigit():
                        print(">> ERROR: Harap masukkan angka!")
                        continue

                    pos = int(pos_input)
                    if 1 <= pos <= count + 1:
                        break
                    else:
                        print(f">> ERROR: Posisi tidak valid! Hanya boleh 1 s/d {count + 1}")
                except ValueError:
                    print(">> ERROR: Harap masukkan angka!")

            # INPUT NIM
            nim = ""
            while True:
                nim = input("Masukkan NIM: ")
                if not nim.isdigit():
                    print(">> ERROR: NIM harus berupa ANGKA saja!")
                    continue

                is_duplicate = False
                for i in range(count):
                    if nim_array[i] == nim:
                        is_duplicate = True
                        break
                if is_duplicate:
                    print(f">> ERROR: NIM {nim} sudah ada!")
                else:
                    break

            # INPUT NAMA
            nama = ""
            while True:
                nama = input("Masukkan Nama: ")
                is_nama_duplicate = False
                for i in range(count):
                    if nama_array[i].lower() == nama.lower():
                        is_nama_duplicate = True
                        break
                if is_nama_duplicate:
                    print(f">> ERROR: Nama \"{nama}\" sudah ada!")
                else:
                    break

            # INPUT JK
            jk = ""
            while True:
                input_jk = input("Masukkan Jenis Kelamin (L/P): ").upper()
                if input_jk == "L" or input_jk == "P":
                    jk = input_jk
                    break
                else:
                    print(">> ERROR: Input salah! Ketik 'L' atau 'P'.")

            # PROSES INSERT POSISI (Geser ke Kanan)
            index = pos - 1
            for i in range(count, index, -1):
                nim_array[i] = nim_array[i-1]
                nama_array[i] = nama_array[i-1]
                jk_array[i] = jk_array[i-1]

            nim_array[index] = nim
            nama_array[index] = nama
            jk_array[index] = jk
            count += 1
            print(f">> SUKSES: Data disisipkan di posisi {pos}")

    # 3. INSERT AT END
    elif choice == 3:
        if count >= capacity:
            print(">> GAGAL: Array Penuh!")
        else:
            # INPUT NIM
            nim = ""
            while True:
                nim = input("Masukkan NIM: ")
                if not nim.isdigit():
                    print(">> ERROR: NIM harus berupa ANGKA saja!")
                    continue
                is_duplicate = False
                for i in range(count):
                    if nim_array[i] == nim:
                        is_duplicate = True
                        break
                if is_duplicate:
                    print(f">> ERROR: NIM {nim} sudah ada!")
                else:
                    break

            # INPUT NAMA
            nama = ""
            while True:
                nama = input("Masukkan Nama: ")
                is_nama_duplicate = False
                for i in range(count):
                    if nama_array[i].lower() == nama.lower():
                        is_nama_duplicate = True
                        break
                if is_nama_duplicate:
                    print(f">> ERROR: Nama \"{nama}\" sudah ada!")
                else:
                    break

            # INPUT JK
            jk = ""
            while True:
                input_jk = input("Masukkan Jenis Kelamin (L/P): ").upper()
                if input_jk == "L" or input_jk == "P":
                    jk = input_jk
                    break
                else:
                    print(">> ERROR: Input salah! Ketik 'L' atau 'P'.")

            # PROSES INSERT BELAKANG
            nim_array[count] = nim
            nama_array[count] = nama
            jk_array[count] = jk
            count += 1
            print(">> SUKSES: Data ditambahkan di akhir.")

    # 4. DELETE FROM BEGINNING
    elif choice == 4:
        if count == 0:
            print(">> GAGAL: Array Kosong!")
        else:
            # Geser ke Kiri (Maju)
            for i in range(0, count - 1):
                nim_array[i] = nim_array[i+1]
                nama_array[i] = nama_array[i+1]
                jk_array[i] = jk_array[i+1]

            # Hapus data terakhir (bersihkan sampah)
            nim_array[count - 1] = None
            nama_array[count - 1] = None
            jk_array[count - 1] = None
            count -= 1
            print(">> SUKSES: Data pertama dihapus.")

    # 5. DELETE GIVEN POSITION
    elif choice == 5:
        if count == 0:
            print(">> GAGAL: Array Kosong!")
        else:
            pos = 0
            while True:
                try:
                    pos_input = input(f"Masukkan Posisi Hapus (1 - {count}): ")
                    if not pos_input.isdigit():
                        print(">> ERROR: Harap masukkan angka!")
                        continue

                    pos = int(pos_input)
                    if 1 <= pos <= count:
                        break
                    else:
                        print(f">> ERROR: Posisi tidak ditemukan! Pilih 1 s/d {count}")
                except ValueError:
                    print(">> ERROR: Harap masukkan angka!")

            index = pos - 1
            # Geser ke Kiri
            for i in range(index, count - 1):
                nim_array[i] = nim_array[i+1]
                nama_array[i] = nama_array[i+1]
                jk_array[i] = jk_array[i+1]

            nim_array[count - 1] = None
            nama_array[count - 1] = None
            jk_array[count - 1] = None
            count -= 1
            print(f">> SUKSES: Data posisi {pos} dihapus.")

    # 6. DELETE FROM END
    elif choice == 6:
        if count == 0:
            print(">> GAGAL: Array Kosong!")
        else:
            nim_array[count - 1] = None
            nama_array[count - 1] = None
            jk_array[count - 1] = None
            count -= 1
            print(">> SUKSES: Data terakhir dihapus.")

    # 7. DELETE BY NIM (FIRST OCCURRENCE)
    elif choice == 7:
        if count == 0:
            print(">> GAGAL: Array Kosong!")
        else:
            target_nim = input("Masukkan NIM yang akan dihapus: ")
            found_index = -1

            # Linear Search
            for i in range(count):
                if nim_array[i] == target_nim:
                    found_index = i
                    break

            if found_index != -1:
                # Geser Kiri
                for i in range(found_index, count - 1):
                    nim_array[i] = nim_array[i+1]
                    nama_array[i] = nama_array[i+1]
                    jk_array[i] = jk_array[i+1]

                nim_array[count - 1] = None
                nama_array[count - 1] = None
                jk_array[count - 1] = None
                count -= 1
                print(f">> SUKSES: NIM {target_nim} dihapus.")
            else:
                print(">> GAGAL: NIM tidak ditemukan.")

    # 8. SHOW DATA
    elif choice == 8:
        print("\nData Mahasiswa:")
        print("+-----+--------------+---------------------------+-----+")
        # Menggunakan f-string formatting untuk merapikan tabel
        # :<3 artinya rata kiri, lebar 3
        # :<12 artinya rata kiri, lebar 12, dst
        print(f"| {'No':<3} | {'NIM':<12} | {'Nama':<25} | {'L/P':<3} |")
        print("+-----+--------------+---------------------------+-----+")

        if count == 0:
            print(f"| {'DATA KOSONG':^50} |") # ^50 artinya rata tengah lebar 50
        else:
            for i in range(count):
                print(f"| {i+1:<3} | {nim_array[i]:<12} | {nama_array[i]:<25} | {jk_array[i]:<3} |")

        print("+-----+--------------+---------------------------+-----+")

    # 9. EXIT
    elif choice == 9:
        print("Keluar program...")
        break

    else:
        print(">> Menu tidak valid!")

File: tugasFariz.java
public class tugasFariz {
    public static void main(String[] args) {

        String[][] dataArray = new String[10][3];

        int count = 0;
        int capacity = 10;
        int choice;

        java.util.Scanner sc = new java.util.Scanner(System.in);

        // HEADER
        System.out.println("\n=================================================");
        System.out.println("            TUGAS STRUKTUR DATA ARRAY            ");
        System.out.println("=================================================");
        System.out.println(" Nama    : Fariz Husain Albar");
        System.out.println(" Kelas   : Informatika A");
        System.out.println(" NIM     : 24106050011");
        System.out.println(" Matkul  : Struktur Data");
        System.out.println("=================================================");
        System.out.println(" >> Tekan ENTER untuk masuk ke Menu Program...");
        sc.nextLine();

        while (true) {
            System.out.println("\n=== MENU MAHASISWA (Count: " + count + "/" + capacity + ") ===");
            System.out.println("1. Insert at beginning");
            System.out.println("2. Insert at given position");
            System.out.println("3. Insert at end");
            System.out.println("4. Delete from beginning");
            System.out.println("5. Delete given position");
            System.out.println("6. Delete from end");
            System.out.println("7. Delete first occurence (by NIM)");
            System.out.println("8. Show data");
            System.out.println("9. Exit");
            System.out.print("Pilih menu: ");

            if (sc.hasNextInt()) {
                choice = sc.nextInt();
                sc.nextLine();
            } else {
                System.out.println(">> Input salah! Harap masukkan angka.");
                sc.next();
                continue;
            }

            // LOGIKA MENU

            // 1. INSERT AT BEGINNING
            if (choice == 1) {
                if (count >= capacity) {
                    System.out.println(">> GAGAL: Array Penuh!");
                } else {
                    // INPUT NIM
                    String nim = "";
                    while (true) {
                        System.out.print("Masukkan NIM: ");
                        nim = sc.nextLine();

                        if (!nim.matches("[0-9]+")) {
                            System.out.println(">> ERROR: NIM harus berupa ANGKA saja! Coba lagi.");
                            continue;
                        }

                        boolean isDuplicate = false;
                        for (int i = 0; i < count; i++) {
                            // Cek Kolom 0 (NIM)
                            if (dataArray[i][0].equals(nim)) {
                                isDuplicate = true;
                                break;
                            }
                        }

                        if (isDuplicate) {
                            System.out.println(">> ERROR: NIM " + nim + " sudah ada! Coba NIM lain.");
                        } else {
                            break;
                        }
                    }

                    // INPUT NAMA
                    String nama = "";
                    while (true) {
                        System.out.print("Masukkan Nama: ");
                        nama = sc.nextLine();

                        boolean isNamaDuplicate = false;
                        for (int i = 0; i < count; i++) {
                            // Cek Kolom 1 (Nama)
                            if (dataArray[i][1].equalsIgnoreCase(nama)) {
                                isNamaDuplicate = true;
                                break;
                            }
                        }

                        if (isNamaDuplicate) {
                            System.out.println(">> ERROR: Nama \"" + nama + "\" sudah ada! Coba nama lain.");
                        } else {
                            break;
                        }
                    }

                    // INPUT JK
                    String jk = "";
                    while (true) {
                        System.out.print("Masukkan Jenis Kelamin (L/P): ");
                        String inputJk = sc.nextLine().toUpperCase();
                        if (inputJk.equals("L") || inputJk.equals("P")) {
                            jk = inputJk;
                            break;
                        } else {
                            System.out.println(">> ERROR: Input salah! Ketik 'L' atau 'P'.");
                        }
                    }

                    // PROSES INSERT DEPAN
                    for (int i = count; i > 0; i--) {
                        dataArray[i][0] = dataArray[i - 1][0]; // NIM
                        dataArray[i][1] = dataArray[i - 1][1]; // Nama
                        dataArray[i][2] = dataArray[i - 1][2]; // JK
                    }
                    dataArray[0][0] = nim;
                    dataArray[0][1] = nama;
                    dataArray[0][2] = jk;
                    count++;
                    System.out.println(">> SUKSES: Data disimpan di awal.");
                }
            }

            // 2. INSERT AT GIVEN POSITION
            else if (choice == 2) {
                if (count >= capacity) {
                    System.out.println(">> GAGAL: Array Penuh!");
                } else {
                    // INPUT POSISI
                    int pos = 0;
                    while (true) {
                        System.out.print("Masukkan Posisi Insert (1 - " + (count + 1) + "): ");
                        if (sc.hasNextInt()) {
                            pos = sc.nextInt();
                            sc.nextLine();

                            if (pos >= 1 && pos <= count + 1) {
                                break;
                            } else {
                                System.out.println(">> ERROR: Posisi tidak valid! Hanya boleh 1 s/d " + (count + 1));
                            }
                        } else {
                            System.out.println(">> ERROR: Harap masukkan angka!");
                            sc.next();
                        }
                    }

                    // INPUT NIM
                    String nim = "";
                    while (true) {
                        System.out.print("Masukkan NIM: ");
                        nim = sc.nextLine();

                        if (!nim.matches("[0-9]+")) {
                            System.out.println(">> ERROR: NIM harus berupa ANGKA saja!");
                            continue;
                        }

                        boolean isDuplicate = false;
                        for (int i = 0; i < count; i++) {
                            if (dataArray[i][0].equals(nim)) {
                                isDuplicate = true;
                                break;
                            }
                        }
                        if (isDuplicate) {
                            System.out.println(">> ERROR: NIM " + nim + " sudah ada!");
                        } else {
                            break;
                        }
                    }

                    // INPUT NAMA
                    String nama = "";
                    while (true) {
                        System.out.print("Masukkan Nama: ");
                        nama = sc.nextLine();

                        boolean isNamaDuplicate = false;
                        for (int i = 0; i < count; i++) {
                            if (dataArray[i][1].equalsIgnoreCase(nama)) {
                                isNamaDuplicate = true;
                                break;
                            }
                        }
                        if (isNamaDuplicate) {
                            System.out.println(">> ERROR: Nama \"" + nama + "\" sudah ada!");
                        } else {
                            break;
                        }
                    }

                    // INPUT JK
                    String jk = "";
                    while (true) {
                        System.out.print("Masukkan Jenis Kelamin (L/P): ");
                        String inputJk = sc.nextLine().toUpperCase();
                        if (inputJk.equals("L") || inputJk.equals("P")) {
                            jk = inputJk;
                            break;
                        } else {
                            System.out.println(">> ERROR: Input salah! Ketik 'L' atau 'P'.");
                        }
                    }

                    // PROSES INSERT POSISI
                    int index = pos - 1;
                    for (int i = count; i > index; i--) {
                        dataArray[i][0] = dataArray[i - 1][0];
                        dataArray[i][1] = dataArray[i - 1][1];
                        dataArray[i][2] = dataArray[i - 1][2];
                    }
                    dataArray[index][0] = nim;
                    dataArray[index][1] = nama;
                    dataArray[index][2] = jk;
                    count++;
                    System.out.println(">> SUKSES: Data disisipkan di posisi " + pos);
                }
            }

            // 3. INSERT AT END
            else if (choice == 3) {
                if (count >= capacity) {
                    System.out.println(">> GAGAL: Array Penuh!");
                } else {
                    // INPUT NIM
                    String nim = "";
                    while (true) {
                        System.out.print("Masukkan NIM: ");
                        nim = sc.nextLine();

                        if (!nim.matches("[0-9]+")) {
                            System.out.println(">> ERROR: NIM harus berupa ANGKA saja!");
                            continue;
                        }

                        boolean isDuplicate = false;
                        for (int i = 0; i < count; i++) {
                            if (dataArray[i][0].equals(nim)) {
                                isDuplicate = true;
                                break;
                            }
                        }
                        if (isDuplicate) {
                            System.out.println(">> ERROR: NIM " + nim + " sudah ada!");
                        } else {
                            break;
                        }
                    }

                    // INPUT NAMA
                    String nama = "";
                    while (true) {
                        System.out.print("Masukkan Nama: ");
                        nama = sc.nextLine();

                        boolean isNamaDuplicate = false;
                        for (int i = 0; i < count; i++) {
                            if (dataArray[i][1].equalsIgnoreCase(nama)) {
                                isNamaDuplicate = true;
                                break;
                            }
                        }
                        if (isNamaDuplicate) {
                            System.out.println(">> ERROR: Nama \"" + nama + "\" sudah ada!");
                        } else {
                            break;
                        }
                    }

                    // INPUT JK
                    String jk = "";
                    while (true) {
                        System.out.print("Masukkan Jenis Kelamin (L/P): ");
                        String inputJk = sc.nextLine().toUpperCase();
                        if (inputJk.equals("L") || inputJk.equals("P")) {
                            jk = inputJk;
                            break;
                        } else {
                            System.out.println(">> ERROR: Input salah! Ketik 'L' atau 'P'.");
                        }
                    }

                    // PROSES INSERT BELAKANG
                    dataArray[count][0] = nim;
                    dataArray[count][1] = nama;
                    dataArray[count][2] = jk;
                    count++;
                    System.out.println(">> SUKSES: Data ditambahkan di akhir.");
                }
            }

            // 4. DELETE FROM BEGINNING
            else if (choice == 4) {
                if (count == 0) {
                    System.out.println(">> GAGAL: Array Kosong!");
                } else {
                    for (int i = 0; i < count - 1; i++) {
                        dataArray[i][0] = dataArray[i + 1][0];
                        dataArray[i][1] = dataArray[i + 1][1];
                        dataArray[i][2] = dataArray[i + 1][2];
                    }
                    dataArray[count - 1][0] = null;
                    dataArray[count - 1][1] = null;
                    dataArray[count - 1][2] = null;
                    count--;
                    System.out.println(">> SUKSES: Data pertama dihapus.");
                }
            }

            // 5. DELETE GIVEN POSITION
            else if (choice == 5) {
                if (count == 0) {
                    System.out.println(">> GAGAL: Array Kosong!");
                } else {
                    // INPUT POSISI HAPUS
                    int pos = 0;
                    while (true) {
                        System.out.print("Masukkan Posisi Hapus (1 - " + count + "): ");
                        if (sc.hasNextInt()) {
                            pos = sc.nextInt();
                            sc.nextLine();

                            if (pos >= 1 && pos <= count) {
                                break;
                            } else {
                                System.out.println(">> ERROR: Posisi tidak ditemukan! Pilih 1 s/d " + count);
                            }
                        } else {
                            System.out.println(">> ERROR: Harap masukkan angka!");
                            sc.next();
                        }
                    }

                    // PROSES HAPUS
                    int index = pos - 1;
                    for (int i = index; i < count - 1; i++) {
                        dataArray[i][0] = dataArray[i + 1][0];
                        dataArray[i][1] = dataArray[i + 1][1];
                        dataArray[i][2] = dataArray[i + 1][2];
                    }
                    dataArray[count - 1][0] = null;
                    dataArray[count - 1][1] = null;
                    dataArray[count - 1][2] = null;
                    count--;
                    System.out.println(">> SUKSES: Data posisi " + pos + " dihapus.");
                }
            }

            // 6. DELETE FROM END
            else if (choice == 6) {
                if (count == 0) {
                    System.out.println(">> GAGAL: Array Kosong!");
                } else {
                    dataArray[count - 1][0] = null;
                    dataArray[count - 1][1] = null;
                    dataArray[count - 1][2] = null;
                    count--;
                    System.out.println(">> SUKSES: Data terakhir dihapus.");
                }
            }

            // 7. DELETE BY NIM (FIRST OCCURRENCE)
            else if (choice == 7) {
                if (count == 0) {
                    System.out.println(">> GAGAL: Array Kosong!");
                } else {
                    System.out.print("Masukkan NIM yang akan dihapus: ");
                    String targetNim = sc.nextLine();
                    int foundIndex = -1;

                    for (int i = 0; i < count; i++) {
                        // Cek Kolom 0 (NIM)
                        if (dataArray[i][0].equals(targetNim)) {
                            foundIndex = i;
                            break;
                        }
                    }

                    if (foundIndex != -1) {
                        for (int i = foundIndex; i < count - 1; i++) {
                            dataArray[i][0] = dataArray[i + 1][0];
                            dataArray[i][1] = dataArray[i + 1][1];
                            dataArray[i][2] = dataArray[i + 1][2];
                        }
                        dataArray[count - 1][0] = null;
                        dataArray[count - 1][1] = null;
                        dataArray[count - 1][2] = null;
                        count--;
                        System.out.println(">> SUKSES: NIM " + targetNim + " dihapus.");
                    } else {
                        System.out.println(">> GAGAL: NIM tidak ditemukan.");
                    }
                }
            }

            // 8. SHOW DATA
            else if (choice == 8) {
                System.out.println("\nData Mahasiswa:");
                System.out.println("+-----+--------------+---------------------------+-----+");
                System.out.printf("| %-3s | %-12s | %-25s | %-3s |%n", "No", "NIM", "Nama", "L/P");
                System.out.println("+-----+--------------+---------------------------+-----+");

                if (count == 0) {
                    System.out.printf("| %-50s |%n", "                  DATA KOSONG");
                } else {
                    for (int i = 0; i < count; i++) {
                        // Akses array 2D: [Baris][0], [Baris][1], [Baris][2]
                        System.out.printf("| %-3d | %-12s | %-25s | %-3s |%n",
                                (i + 1), dataArray[i][0], dataArray[i][1], dataArray[i][2]);
                    }
                }
                System.out.println("+-----+--------------+---------------------------+-----+");
            }

            // 9. EXIT PROGRAM
            else if (choice == 9) {
                System.out.println("Keluar program...");
                break;
            } else {
                System.out.println(">> Menu tidak valid!");
            }
        } // End While

        sc.close();
    }
}
